LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -88,46 +85,39 @@ private bool EnsureFirstMessageInCache(string key, LoRaRequest loRaRequest, out
             return false;
         }
 
-        internal static string CreateCacheKey(LoRaPayloadData payload)
+        internal static string CreateCacheKey(LoRaPayloadData payload, LoRaDevice loRaDevice)
         {
-            var totalBufferLength = DevAddr.Size + Mic.Size + (payload.RawMessage?.Length ?? 0) + payload.Fcnt.Length;
-            var buffer = totalBufferLength <= 128 ? stackalloc byte[totalBufferLength] : new byte[totalBufferLength]; // uses the stack for small allocations, otherwise the heap
-
-            var index = 0;
-            _ = payload.DevAddr.Write(buffer);
-            index += DevAddr.Size;
-
-            _ = payload.Mic is { } someMic ? someMic.Write(buffer[index..]) : throw new InvalidOperationException("Mic must not be null.");
-            index += Mic.Size;
-
-            payload.RawMessage?.CopyTo(buffer[index..]);
-            index += payload.RawMessage?.Length ?? 0;
-
-            BinaryPrimitives.WriteUInt16LittleEndian(buffer[index..], BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
+            var totalBufferLength = DevEui.Size + Mic.Size + payload.Fcnt.Length;
+            Span<byte> buffer = stackalloc byte[totalBufferLength];
+            var head = buffer; // keeps a view pointing at the start of the buffer
 
-            var key = Sha256.ComputeHash(buffer.ToArray());
+            buffer = DevEui.Parse(loRaDevice.DevEUI.AsSpan()).Write(buffer);
+            buffer = payload.Mic is { } someMic ? someMic.Write(buffer) : throw new InvalidOperationException("Mic must not be null.");
```

Comment:
The fact that the payload must have a MIC should be valid at the beginning of the method and `payload` identified as an invalid argument, as shown below:

```diff
diff --git a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs
index 2e4a08cf..4e155ec0 100644
--- a/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs
+++ b/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs
@@ -87,13 +86,14 @@ namespace LoRaWan.NetworkServer
 
         internal static string CreateCacheKey(LoRaPayloadData payload, LoRaDevice loRaDevice)
         {
+            var mic = payload.Mic ?? throw new ArgumentException("Payload MIC must be defined.");
             var totalBufferLength = DevEui.Size + Mic.Size + payload.Fcnt.Length;
             Span<byte> buffer = stackalloc byte[totalBufferLength];
             var head = buffer; // keeps a view pointing at the start of the buffer
 
             buffer = DevEui.Parse(loRaDevice.DevEUI.AsSpan()).Write(buffer);
-            buffer = payload.Mic is { } someMic ? someMic.Write(buffer) : throw new InvalidOperationException("Mic must not be null.");
+            buffer = mic.Write(buffer);
             BinaryPrimitives.WriteUInt16LittleEndian(buffer, BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
 
             return CreateCacheKeyCore(DataMessageCacheKeyPrefix, head);
         }
```

Reply:
Fixed in a7594688cac18a80e93e6dd92dbd0d01db0788e7